File: backend/app/analyzer/engine.py
"""Moteur principal : interface abstraite + implémentation déterministe.

``Analyzer`` est l'interface commune : une future implémentation
``AIAnalyzer`` (modèle externe) pourra s'insérer sans modifier le reste
du pipeline. ``RuleBasedAnalyzer`` fonctionne SANS aucune API d'IA :
expressions régulières, normalisation, dictionnaires, scoring et
correspondance avec le catalogue de titres.
"""

from __future__ import annotations

import logging
from abc import ABC, abstractmethod
from dataclasses import dataclass

from .aliases import AliasRegistry
from .extractors import TextAnalysis, analyze_text
from .models import MEDIA_TYPES, AnalysisResult
from .quality import infer_from_resolution, rank_of
from .scoring import score_analysis
from .text_utils import normalize, strip_extension

logger = logging.getLogger("animebox.analyzer")

AUDIO_EXTENSIONS = {".mp3", ".m4a", ".aac", ".flac", ".ogg", ".opus", ".wav"}
VIDEO_EXTENSIONS = {".mkv", ".mp4", ".avi", ".webm", ".mov", ".m4v", ".ts", ".m2ts"}
IMAGE_EXTENSIONS = {".jpg", ".jpeg", ".png", ".webp", ".gif"}
SUBTITLE_EXTENSIONS = {".srt", ".ass", ".ssa", ".vtt"}


@dataclass
class EngineConfig:
    """Stratégie de priorité des informations (configurable).

    Ordre par défaut (du plus fiable au moins fiable) :
        1. métadonnées du fichier (durée, dimensions…)  → gérées à part
        2. nom du fichier
        3. caption Telegram
        4. texte du message
    """

    text_field_order: tuple[str, ...] = ("file_name", "text")


class Analyzer(ABC):
    """Interface abstraite du moteur d'analyse."""

    name = "abstract"

    @abstractmethod
    def analyze(self, message: dict, context: dict | None = None) -> AnalysisResult:
        """Analyse une publication Telegram (dict) et renvoie un résultat structuré."""


class RuleBasedAnalyzer(Analyzer):
    """Implémentation déterministe (aucune API d'IA externe)."""

    name = "rule-based"
    version = "1.0.0"

    def __init__(self, registry: AliasRegistry | None = None, config: EngineConfig | None = None):
        self.registry = registry or AliasRegistry()
        self.config = config or EngineConfig()

    # ------------------------------------------------------------------ API

    def analyze(self, message: dict, context: dict | None = None) -> AnalysisResult:
        message = {str(key): value for key, value in (message or {}).items()}

        # 1. Sources de texte, dans l'ordre de priorité configuré.
        file_name = strip_extension(message.get("file_name"))
        text = message.get("text") or message.get("caption")
        sources: list[TextAnalysis] = []
        for label in self.config.text_field_order:
            raw = file_name if label == "file_name" else text
            if raw and str(raw).strip():
                sources.append(analyze_text(normalize(raw), label, self.registry))

        # 2. Fusion des sources (la plus prioritaire l'emporte par champ).
        result = self._merge(sources, message)

        # 3. Qualité déduite des métadonnées du fichier si aucune valeur explicite.
        width = _as_int(message.get("width"))
        height = _as_int(message.get("height"))
        if result.quality is None and (width or height):
            inferred = infer_from_resolution(width, height)
            if inferred is not None:
                result.quality = inferred
                result.quality_original = f"{width}x{height}"
                result.quality_source = "metadata"
                result.quality_rank = rank_of(inferred)

        # 4. Informations du fichier (jamais inventées : None si absentes).
        result.file_name = message.get("file_name")
        result.file_size = _as_int(message.get("file_size"))
        result.mime_type = message.get("mime_type")
        result.duration = _as_int(message.get("duration"))
        result.width = width
        result.height = height

        # 5. Type de média.
        result.media_type = self._media_type(message)

        # 6. Références Telegram (conservées pour remonter à la publication).
        result.telegram_channel_id = message.get("telegram_channel_id") or message.get("channel_id")
        result.telegram_channel_username = message.get("telegram_channel_username") or message.get("channel_username")
        result.telegram_message_id = _as_int(message.get("telegram_message_id") or message.get("message_id"))
        result.telegram_message_link = message.get("telegram_message_link") or message.get("link")

        # 7. Score de confiance et statut.
        result.confidence, result.status = score_analysis(result)

        logger.info(
            "[Analyzer] %s | %s | S%s E%s | %s | %s | %s%% (%s)",
            result.telegram_message_id if result.telegram_message_id is not None else "-",
            result.title or "?",
            result.season if result.season is not None else "?",
            result.episode if result.episode is not None else "?",
            result.quality or "?",
            result.language or "?",
            result.confidence,
            result.status,
        )
        return result

    # ------------------------------------------------------------- internes

    def _merge(self, sources: list[TextAnalysis], message: dict) -> AnalysisResult:
        result = AnalysisResult()

        def first_value(attribute: str):
            """Première source (ordre de priorité) qui renseigne l'attribut."""
            for source in sources:
                value = getattr(source.extraction, attribute)
                if value is not None:
                    return value, source
            return None, None

        for attribute, target, source_target in (
            ("season", "season", "season_source"),
            ("year", "year", None),
        ):
            value, source = first_value(attribute)
            setattr(result, target, value)
            if source_target and value is not None:
                setattr(result, source_target, getattr(source.extraction, source_target))

        # Épisode (numéro + type « spécial »).
        value, source = first_value("episode")
        result.episode = value
        result.episode_source = getattr(source.extraction, "episode_source", None) if value is not None else None
        value, source = first_value("episode_kind")
        result.episode_kind = value
        if result.episode is None and result.episode_kind is not None:
            result.episode_source = getattr(source.extraction, "episode_source", None)

        # Qualité explicite (la fusion garde la valeur de la source prioritaire).
        value, source = first_value("quality")
        result.quality = value
        if value is not None:
            result.quality_original = source.extraction.quality_original
            result.quality_source = "explicit"
            result.quality_rank = rank_of(value)

        # Langue et sous-titres : champs indépendants.
        value, source = first_value("language")
        result.language = value or "unknown"
        value, source = first_value("subtitles")
        result.subtitles = value

        # Titre : on préfère la source dont le titre est RECONNU au catalogue
        # (une caption complète l'emporte sur un nom de fichier tronqué),
        # sinon la source la plus prioritaire qui propose un titre.
        chosen = next(
            (source for source in sources if source.title.key and source.title.matched),
            None,
        )
        if chosen is None:
            chosen = next((source for source in sources if source.title.key), None)
        if chosen is not None:
            result.title = chosen.title.display
            result.title_key = chosen.title.key
            result.title_matched = chosen.title.matched
            result.title_via_alias = chosen.title.via_alias
            if chosen.title.anime is not None:
                result.anime_key = chosen.title.anime.key
                result.original_title = chosen.title.anime.original_title
                result.release_year = chosen.title.anime.release_year

        # Contradictions entre sources + avertissements cumulés.
        for field in ("season", "episode", "quality", "language", "year"):
            values = {
                getattr(source.extraction, field)
                for source in sources
                if getattr(source.extraction, field) is not None
            }
            if len(values) > 1:
                result.warnings.append(f"{field} contradictoire entre les sources : {sorted(str(v) for v in values)}")
        for source in sources:
            for warning in source.extraction.warnings:
                if warning not in result.warnings:
                    result.warnings.append(warning)
        return result

    @staticmethod
    def _media_type(message: dict) -> str:
        media = str(message.get("media_type") or "unknown").lower()
        if media not in MEDIA_TYPES:
            media = "unknown"
        file_name = str(message.get("file_name") or "").lower()
        mime = str(message.get("mime_type") or "").lower()
        extension = f".{file_name.rsplit('.', 1)[-1]}" if "." in file_name else ""
        if mime.startswith("audio/") or extension in AUDIO_EXTENSIONS:
            return "audio"
        if mime.startswith("video/") or extension in VIDEO_EXTENSIONS:
            return "video"
        if mime.startswith("image/") or extension in IMAGE_EXTENSIONS:
            return "image"
        if extension in SUBTITLE_EXTENSIONS:
            return "document"  # fichier de sous-titres
        if media == "video" and mime:
            return "video"
        return media


def _as_int(value) -> int | None:
    try:
        return int(value)
    except (TypeError, ValueError):
        return None

File: lib/features/analyzer/engine.dart
/// Moteur principal local — port fidèle de `backend/app/analyzer/engine.py`.
///
/// Analyse déterministe (aucune API externe) : normalisation, motifs,
/// dictionnaires, scoring et correspondance avec le catalogue de titres.
library;

import 'aliases.dart';
import 'extractors.dart';
import 'models.dart';
import 'quality.dart' as quality;
import 'scoring.dart';
import 'text_utils.dart';

const Set<String> audioExtensions = {'.mp3', '.m4a', '.aac', '.flac', '.ogg', '.opus', '.wav'};
const Set<String> videoExtensions = {'.mkv', '.mp4', '.avi', '.webm', '.mov', '.m4v', '.ts', '.m2ts'};
const Set<String> imageExtensions = {'.jpg', '.jpeg', '.png', '.webp', '.gif'};
const Set<String> subtitleExtensions = {'.srt', '.ass', '.ssa', '.vtt'};
const Set<String> mediaTypes = {'video', 'document', 'image', 'audio', 'text', 'unknown'};

/// Stratégie de priorité des informations (configurable).
class EngineConfig {
  const EngineConfig({this.textFieldOrder = const ['file_name', 'text']});

  /// Ordre du plus fiable au moins fiable : nom de fichier, puis texte.
  final List<String> textFieldOrder;
}

int? asInt(Object? value) {
  if (value is int) return value;
  if (value is num) return value.toInt();
  if (value is String) return int.tryParse(value);
  return null;
}

/// Analyse une publication Telegram (map) et renvoie un résultat structuré.
///
/// La map est volontairement identique au format du moteur Python :
/// `file_name`, `text`/`caption`, `file_size`, `mime_type`, `duration`,
/// `width`, `height`, `telegram_channel_id`/`channel_id`,
/// `telegram_channel_username`/`channel_username`,
/// `telegram_message_id`/`message_id`, `telegram_message_link`/`link`,
/// `media_type`.
class RuleBasedAnalyzer {
  RuleBasedAnalyzer({AliasRegistry? registry, EngineConfig? config})
      : registry = registry ?? AliasRegistry(),
        config = config ?? const EngineConfig();

  final AliasRegistry registry;
  final EngineConfig config;

  AnalysisResult analyze(Map<String, Object?> message) {
    // 1. Sources de texte, dans l'ordre de priorité configuré.
    final String? fileName = stripExtension(message['file_name']?.toString());
    final Object? rawText = message['text'] ?? message['caption'];
    final List<TextAnalysis> sources = [];
    for (final String label in config.textFieldOrder) {
      final String? raw = label == 'file_name' ? fileName : rawText?.toString();
      if (raw != null && raw.trim().isNotEmpty) {
        sources.add(analyzeText(normalize(raw), label, registry));
      }
    }

    // 2. Fusion des sources (la plus prioritaire l'emporte par champ).
    final AnalysisResult result = _merge(sources, message);

    // 3. Qualité déduite des métadonnées du fichier si aucune valeur explicite.
    final int? width = asInt(message['width']);
    final int? height = asInt(message['height']);
    if (result.quality == null && (width != null || height != null)) {
      final String? inferred = quality.inferFromResolution(width, height);
      if (inferred != null) {
        result.quality = inferred;
        result.qualityOriginal = '$width${height == null ? '' : 'x$height'}';
        result.qualitySource = 'metadata';
        result.qualityRank = quality.rankOf(inferred);
      }
    }

    // 4. Informations du fichier (jamais inventées : null si absentes).
    result.fileName = message['file_name']?.toString();
    result.fileSize = asInt(message['file_size']);
    result.mimeType = message['mime_type']?.toString();
    result.duration = asInt(message['duration']);
    result.width = width;
    result.height = height;

    // 5. Type de média.
    result.mediaType = _mediaType(message);

    // 6. Références Telegram (conservées pour remonter à la publication).
    result.telegramChannelId =
        (message['telegram_channel_id'] ?? message['channel_id'])?.toString();
    result.telegramChannelUsername =
        (message['telegram_channel_username'] ?? message['channel_username'])?.toString();
    result.telegramMessageId = asInt(message['telegram_message_id'] ?? message['message_id']);
    result.telegramMessageLink =
        (message['telegram_message_link'] ?? message['link'])?.toString();

    // 7. Score de confiance et statut.
    final (int confidence, String status) = scoreAnalysis(result);
    result.confidence = confidence;
    result.status = status;
    return result;
  }

  String _mediaType(Map<String, Object?> message) {
    String media = (message['media_type'] ?? 'unknown').toString().toLowerCase();
    if (!mediaTypes.contains(media)) media = 'unknown';
    final String fileName = (message['file_name'] ?? '').toString().toLowerCase();
    final String mime = (message['mime_type'] ?? '').toString().toLowerCase();
    final String extension = fileName.contains('.') ? '.${fileName.split('.').last}' : '';
    if (mime.startsWith('audio/') || audioExtensions.contains(extension)) return 'audio';
    if (mime.startsWith('video/') || videoExtensions.contains(extension)) return 'video';
    if (mime.startsWith('image/') || imageExtensions.contains(extension)) return 'image';
    if (subtitleExtensions.contains(extension)) return 'document';
    if (media == 'video' && mime.isNotEmpty) return 'video';
    return media;
  }

  AnalysisResult _merge(List<TextAnalysis> sources, Map<String, Object?> message) {
    final AnalysisResult result = AnalysisResult();

    (Object?, TextAnalysis?) firstValue(String attribute) {
      for (final TextAnalysis source in sources) {
        final Object? value = _attributeOf(source.extraction, attribute);
        if (value != null) return (value, source);
      }
      return (null, null);
    }

    // Saison + année.
    final (Object? seasonValue, TextAnalysis? seasonSource) = firstValue('season');
    result.season = seasonValue as int?;
    if (seasonValue != null && seasonSource != null) {
      result.seasonSource = seasonSource.extraction.seasonSource;
    }
    final (Object? yearValue, _) = firstValue('year');
    result.year = yearValue as int?;

    // Épisode (numéro + type « spécial »).
    final (Object? episodeValue, TextAnalysis? episodeSource) = firstValue('episode');
    result.episode = episodeValue as int?;
    if (episodeValue != null && episodeSource != null) {
      result.episodeSource = episodeSource.extraction.episodeSource;
    }
    final (Object? kindValue, TextAnalysis? kindSource) = firstValue('episode_kind');
    result.episodeKind = kindValue as String?;
    if (result.episode == null && result.episodeKind != null && kindSource != null) {
      result.episodeSource = kindSource.extraction.episodeSource;
    }

    // Qualité explicite (la source prioritaire l'emporte).
    final (Object? qualityValue, TextAnalysis? qualitySource) = firstValue('quality');
    result.quality = qualityValue as String?;
    if (qualityValue != null && qualitySource != null) {
      result.qualityOriginal = qualitySource.extraction.qualityOriginal;
      result.qualitySource = 'explicit';
      result.qualityRank = quality.rankOf(result.quality);
    }

    // Langue et sous-titres : champs indépendants.
    final (Object? languageValue, _) = firstValue('language');
    result.language = (languageValue as String?) ?? 'unknown';
    final (Object? subtitlesValue, _) = firstValue('subtitles');
    result.subtitles = subtitlesValue as String?;

    // Titre : on préfère la source dont le titre est RECONNU au catalogue,
    // sinon la source la plus prioritaire qui propose un titre.
    TextAnalysis? chosen;
    for (final TextAnalysis source in sources) {
      if (source.title.key != null && source.title.matched) {
        chosen = source;
        break;
      }
    }
    chosen ??= sources.where((TextAnalysis s) => s.title.key != null).firstOrNull;
    if (chosen != null) {
      result.title = chosen.title.display;
      result.titleKey = chosen.title.key;
      result.titleMatched = chosen.title.matched;
      result.titleViaAlias = chosen.title.viaAlias;
      final AnimeTitle? anime = chosen.title.anime;
      if (anime != null) {
        result.animeKey = anime.key;
        result.originalTitle = anime.originalTitle;
        result.releaseYear = anime.releaseYear;
      }
    }

    // Contradictions entre sources + avertissements cumulés.
    for (final String field in const ['season', 'episode', 'quality', 'language', 'year']) {
      final Set<Object?> values = {
        for (final TextAnalysis source in sources)
          if (_attributeOf(source.extraction, field) != null) _attributeOf(source.extraction, field),
      };
      if (values.length > 1) {
        final List<String> sorted = values.map((Object? v) => v.toString()).toList()..sort();
        result.warnings.add('$field contradictoire entre les sources : $sorted');
      }
    }
    for (final TextAnalysis source in sources) {
      for (final String warning in source.extraction.warnings) {
        if (!result.warnings.contains(warning)) result.warnings.add(warning);
      }
    }
    return result;
  }

  static Object? _attributeOf(Extraction extraction, String attribute) => switch (attribute) {
        'season' => extraction.season,
        'year' => extraction.year,
        'episode' => extraction.episode,
        'episode_kind' => extraction.episodeKind,
        'quality' => extraction.quality,
        'language' => extraction.language,
        'subtitles' => extraction.subtitles,
        _ => null,
      };
}
